### DISCUSSION TITLE
Wasm support one day?

### DISCUSSION THREAD

**Discussion by CharlieTap** [cashapp/zipline#843] [General]
I was wondering whether it will be likely that wasm support would be added in the future? How deeply coupled is the library to QuickJs as a runtime and JS as a compilation target?

If we were to embed a wasm runtime, or Android offered support for running wasm binaries we would potentially have dynamically updatable  Kotlin code running as fast or in some cases faster than it would inside of ART/Dalvik.

This assumes we get wasm32-wasi compliant binaries as part of the new kotlin wasm compiler backend.

**Comment by JakeWharton** [comment]
See https://github.com/cashapp/zipline/issues/717.

> If we were to embed a wasm runtime, or Android offered support for running wasm binaries we would potentially have dynamically updatable Kotlin code running as fast or in some cases faster than it would inside of ART/Dalvik.

This will never happen because you are not allowed to ship something which does JITting of machine code and sets executable pages. We can only interpret.

Since we already ship QuickJS bytecode and not raw JS, we'd just be trading one bytecode interpreter for another. It could actually get slower, but it's unlikely to be dramatically faster.

The largest blocker is that the proposals to WASM itself that Kotlin requires are still in the standardization process and thus are not implemented by any interpreter. Maybe in a year or two the situation will change.

I also don't think it's clear that we need WASI, but we're a long way from making such a decision.

**Comment by CharlieTap** [comment]
Thanks for the quick response!

> This will never happen because you are not allowed to ship something which does JITting of machine code and sets executable pages. We can only interpret.

Ah Is see, but we could embed an interpreter like [wasm3](https://github.com/wasm3/wasm3) I guess?

The reason I looking into this originally was because I've been working on Rust backend that is being compiled to WASM. Given that it sounds like WASM is likely to become the defacto bytecode format for zipline, whats the chance of ingesting different languages? For example if I wanted to create a wasm module from rust and avoid the GC for a certain section of code, this could be awesome alternative to using the usual JNI FFI bindgen ceremony

**Reply by JakeWharton** [reply]
wasm3 is the leading candidate, yes, although it's missing reference types. I just learned about [Web49](https://github.com/FastVM/Web49) though which is a competitor.

Support for loading modules from different languages can be added. We were considering it for JS but decided not to (but not for technical reasons). As to how you talk to that code from Kotlin is a Kotlin problem.

**Reply by crowforkotlin** [reply]
> wasm3 is the leading candidate, yes, although it's missing reference types. I just learned about [Web49](https://github.com/FastVM/Web49) though which is a competitor.Wasm3 确实是目前最受欢迎的候选方案，不过它还不支持引用类型（reference types）。不过我刚刚了解到另一个名为 Web49 的竞争方案。
> 
> Support for loading modules from different languages can be added. We were considering it for JS but decided not to (but not for technical reasons). As to how you talk to that code from Kotlin is a Kotlin problem.我们可以添加从不同语言加载模块的功能。我们原本也考虑为 JavaScript 添加这个功能，但最终决定不这么做（不过这并非出于技术原因）。至于如何用 Kotlin 来访问这些模块中的代码，那属于 Kotlin 语言本身的问题。

https://github.com/crowforkotlin/compose-multiplatform-web-dynamic-load-js-plugin

I've established the host-plugin relationship and dynamic plugin invocation through dynamic JS loading. Since this implementation appears to leave the native codebase unaffected, I believe the solution is viable. My plan is to explore extending it on a Zipline fork branch without relying on a sandbox environment (such as QuickJS or WASM). Zipline plugins are currently implemented using Kotlin/JS to generate the output; to produce a WASM file, Kotlin/WasmJs code would be necessary. If a unified codebase is the goal, then using Kotlin/JS directly should suffice.
<img width="1831" height="601" alt="image" src="https://github.com/user-attachments/assets/5a2e01c6-e3ef-4d25-822c-5808605ade7c" />
